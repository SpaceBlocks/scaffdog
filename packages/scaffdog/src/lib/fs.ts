import { readFileSync, statSync } from 'fs';
import { mkdir, readFile, writeFile } from 'fs/promises';
import type { GlobbyOptions } from 'globby';
import globby from 'globby';

export type FsMkdirOptions = {
  recursive?: boolean;
};

export type FsGlobOptions = GlobbyOptions;

export type FsLibrary = {
  mkdir: (filepath: string, options: FsMkdirOptions) => Promise<void>;
  readFile: (filepath: string) => Promise<string>;
  readFileSync: (filepath: string) => string;
  glob: (
    patterns: string | readonly string[],
    options: FsGlobOptions,
  ) => Promise<string[]>;
  writeFile: (filepath: string, data: string) => Promise<void>;
  fileExists: (filepath: string) => boolean;
  directoryExists: (filepath: string) => boolean;
};

export const createFsLibrary = (): FsLibrary => {
  const fstat = (filepath: string) => {
    try {
      return statSync(filepath);
    } catch (e) {
      return null;
    }
  };

  return {
    mkdir: async (filepath, options) => {
      await mkdir(filepath, options);
    },

    readFile: async (filepath) => {
      return await readFile(filepath, 'utf8');
    },

    readFileSync: (filepath) => {
      return readFileSync(filepath, 'utf8');
    },

    glob: async (patterns, options) => {
      return await globby(patterns, options);
    },

    writeFile: async (filepath, data) => {
      return await writeFile(filepath, data, 'utf8');
    },

    fileExists: (filepath) => {
      return fstat(filepath)?.isFile() ?? false;
    },

    directoryExists: (filepath) => {
      return fstat(filepath)?.isDirectory() ?? false;
    },
  };
};
