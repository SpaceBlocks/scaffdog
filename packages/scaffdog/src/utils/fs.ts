import fs from 'fs';
import { promisify } from 'util';

export const writeFile = promisify(fs.writeFile);
export const readFile = promisify(fs.readFile);
export const mkdir = promisify(fs.mkdir);

const fstat = (filepath: string) => {
  try {
    return fs.statSync(filepath);
  } catch (e) {
    return null;
  }
};

export const fileExists = (filepath: string): boolean =>
  fstat(filepath)?.isFile() ?? false;

export const directoryExists = (filepath: string): boolean =>
  fstat(filepath)?.isDirectory() ?? false;
