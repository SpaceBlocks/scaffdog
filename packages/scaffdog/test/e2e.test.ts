import fs from 'fs';
import path from 'path';
import globby from 'globby';
import strip from 'strip-ansi';
import { afterEach, beforeAll, describe, expect, test } from 'vitest';
import { $ } from 'zx';
import pkg from '../package.json';

$.verbose = false;

const paths = {
  bin: pkg.bin.scaffdog,
};

const clear = () => {
  try {
    fs.rmSync(path.resolve(process.cwd(), 'tmp'), {
      recursive: true,
    });
  } catch (e) {}
};

describe('generate', () => {
  beforeAll(() => {
    clear();
  });

  afterEach(() => {
    clear();
  });

  test.each([
    ['basic', '.', []],
    ['basic', 'override', []],
    ['inputs', '.', ['value:scaffdog', 'choice:B']],
    ['inputs', 'deep/nest', ['value:e2e', 'choice:C']],
    [
      'question',
      'dist',
      [
        'shorthand:input_shorthand',
        'input:input_basic',
        'input_with_initial:input_initial',
        'input_if:input_true',
        'bool:true',
        'bool_with_true:false',
        'bool_with_false:true',
        'bool_if:true',
        'list:A',
        'list_with_initial:B',
        'list_if:C',
        'checkbox:A',
        'checkbox_with_initial:B',
        'checkbox_if:C',
      ],
    ],
    ['vars', '.', ['foo:bar-baz']],
    ['conditional-generate', 'true', ['bool:true']],
    ['conditional-generate', 'false', ['bool:false']],
  ])('%s - %s', async (name, output, answers) => {
    const flags = [
      name,
      '-p',
      'fixtures',
      '-o',
      output,
      ...answers.flatMap((v) => ['-a', v]),
    ];

    const p = await $`node ${paths.bin} generate ${flags}`;
    const out = strip(p.toString());

    const files = await globby('tmp/**/*', {
      onlyFiles: true,
    });

    const results = files.reduce<
      {
        path: string;
        content: string;
      }[]
    >((acc, cur) => {
      acc.push({
        path: cur,
        content: fs.readFileSync(cur, 'utf8'),
      });
      return acc;
    }, []);

    expect(out).toMatchSnapshot();
    expect(results).toMatchSnapshot();
  });
});
