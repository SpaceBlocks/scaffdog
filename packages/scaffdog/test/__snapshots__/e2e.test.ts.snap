// Vitest Snapshot v1

exports[`generate > basic - . 1`] = `
"
ℹ Output destination directory: \\".\\"

───┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ✔ │ File: static.txt
───┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 1 │ static content

───┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ✔ │ File: nest/output.txt
───┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 1 │ basic
 2 │ output
 3 │ output.txt
 4 │ nest/output.txt
 5 │ .txt
 6 │ nest

🐶 Generated 2 files!

     ✔ static.txt
     ✔ nest/output.txt

"
`;

exports[`generate > basic - override 1`] = `
"
ℹ Output destination directory: \\"override\\"

───┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ✔ │ File: static.txt
───┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 1 │ static content

───┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ✔ │ File: nest/output.txt
───┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 1 │ basic
 2 │ output
 3 │ output.txt
 4 │ override/nest/output.txt
 5 │ .txt
 6 │ override/nest

🐶 Generated 2 files!

     ✔ override/static.txt
     ✔ override/nest/output.txt

"
`;

exports[`generate > conditional-generate - false 1`] = `
"
ℹ Output destination directory: \\"false\\"

───┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ✔ │ File: false.txt
───┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 1 │ If bool is false, output.

🐶 Generated 1 file! (1 skipped)

     ✔ false/false.txt
     ⚠ false/true.txt (skipped)

"
`;

exports[`generate > conditional-generate - true 1`] = `
"
ℹ Output destination directory: \\"true\\"

───┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ✔ │ File: true.txt
───┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 1 │ If bool is true, output.

🐶 Generated 1 file! (1 skipped)

     ✔ true/true.txt
     ⚠ true/false.txt (skipped)

"
`;

exports[`generate > inputs - . 1`] = `
"
ℹ Output destination directory: \\"output\\"

───┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ✔ │ File: scaffdog.txt
───┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 1 │ value: scaffdog
 2 │ choice: B
 3 │ path: output/scaffdog.txt

🐶 Generated 1 file!

     ✔ output/scaffdog.txt

"
`;

exports[`generate > inputs - deep/nest 1`] = `
"
ℹ Output destination directory: \\"output/deep/nest\\"

───┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ✔ │ File: e2e.txt
───┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 1 │ value: e2e
 2 │ choice: C
 3 │ path: output/deep/nest/e2e.txt

🐶 Generated 1 file!

     ✔ output/deep/nest/e2e.txt

"
`;

exports[`generate > question - dist 1`] = `
"
ℹ Output destination directory: \\"dist\\"

────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  ✔ │ File: result.txt
────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  1 │ shorthand: input_shorthand
  2 │ input: input_basic
  3 │ input_with_initial: input_initial
  4 │ input_if: input_true
  5 │ input_if_with_initial: default
  6 │ input_if_without_initial: 
  7 │ bool: true
  8 │ bool_with_true: false
  9 │ bool_with_false: true
 10 │ bool_if: true
 11 │ bool_if_with_initial: true
 12 │ bool_if_without_initial: false
 13 │ list: A
 14 │ list_with_initial: B
 15 │ list_if: C
 16 │ list_if_with_initial: B
 17 │ list_if_without_initial: 
 18 │ checkbox: A
 19 │ checkbox_with_initial: B
 20 │ checkbox_if_with_initial: B,C
 21 │ checkbox_if_without_initial: 

🐶 Generated 1 file!

     ✔ dist/result.txt

"
`;

exports[`generate > vars - . 1`] = `
"
ℹ Output destination directory: \\".\\"

───┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ✔ │ File: index.txt
───┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 1 │ raw: scaff-dog
 2 │ foo: ScaffDog

🐶 Generated 1 file!

     ✔ index.txt

"
`;
